ipaserver/install/certs.py

```diff
@@ -574,6 +574,7 @@ def export_pkcs12(self, pkcs12_fname, pkcs12_pwd_fname, nickname=None):
         ])
 
     def create_from_cacert(self):
+        self.nssdb.create_db()
         cacert_fname = paths.IPA_CA_CRT
         if os.path.isfile(self.certdb_fname):
```

Comment:
``if self.exists():`` is a better way to check if the NSSDB already exists.

Reply:
I'm suggesting to (also) fix ``os.path.isfile(self.certdb_fname)``. The property ``self.certdb_fname`` redirects to ``self.nssdb.certdb``. The value ``certdb`` of a ``NSSDatabase`` instance is ``None``, in case the NSS DB doesn't exist yet. I had to change the implementation to support DBM and SQL databases. The attribute `certdb` is filled when either a DB already exists or a new DB is created using ``create_db``.

``os.path.isfile(self.certdb_fname)`` was the old way to check if ``cert8.db`` is present. The new way is ``self.exists()``. It makes use of DBM/SQL auto-detection and checks for the presence of all NSS DB related files (cert, key, secmod).

Reply:
I misread the code above, seeing `os.path.isfile(cacert_fname)` instead of what was actually written.

So, given we're now (as of this commit) *creating* the NSSDB, what is the point of retaining the code for looking up the cert.  The NSSDB will be empty.

So can't we just create the NSSDB and import the CA cert?  Or am I missing something...

Reply:
In general it looks like a good idea to always create a DB here -- but only if the DB does not exist yet. The ``create_db`` method replaces any previous DB. The method deletes or renames all DB related files.

Therefore the ``create_db`` call must be protected by a ``if not self.exists()`` call.

Reply:
Right.  I agree.  @rcritten please guard the `create_db` in an `if not self.exists(): ...`.
Then the `if os.path.isfile(self.certdb_fname):` conditional can be removed and the
code in that block moved out one level.

Aside from that, I've tested the code and it does what it says on the tin :)

Reply:
This is never expected to work on an existing database, the name after all is create_from_cacert(). The idea is you are given a CA cert and you end up with an NSS database containing only that cert.

This method is only used once as well, in this one DL0 case.

Still, safety first. I can add a guard and a doc describing how this method is used.